### ISSUE TITLE
How to work on new data?

### ISSUE THREAD

**Issue by LeeJuly30** [kingofspace0wzz/wae-rnf-lm#2]
Very nice work and thanks for sharing, but i have a few questions.
In your paper, in order to maximize the regularized jacobian  we need to calculate the center of posterior codes z. In your code the centers was pretrained and we just load it at the beginning. 
Since I want to work on some new data, i modify ``NormalizingFlows``'s ``forward`` function like this:
```python
def forward(self, z, centers=None):
        log_det_jacobian = []
        penalty = []
        if centers is None:
            with torch.no_grad():
                _, centers = lloyd(z, 20)

        for flow in self.flows:
            z, j, d = flow(z, centers, reg=self.reg, band=self.band)
            log_det_jacobian.append(j)
            penalty.append(d)

        return z, sum(log_det_jacobian), sum(penalty)
```
but the training process become much slower than before, i'm wondering if you have some idea to work on new data? Do i need to firstly train a VAE to calculate the center of posterior codes z and then load it at the beginning of training a wae-rnf? 
Thanks for you help.

**Comment by kingofspace0wzz** [comment]
Hi @LeeJuly30, thanks for your interests!

In your code, my first impression is that you are dynamically changing the center for different latent variable z. This is incorrect. The clusters/centers are a discrete approximation of some first-hand initial manifold, which should be fixed during training. The relative training time would be the same since you don't need to do an extra lloyds clustering in line 4-6.

Different datasets have distinct first-hand manifolds to work with at the beginning. So yes, you should pretrain a VAE model (with KL-annealing), and then do clustering on the latent variable z sampling from the pretrained model. The number of centers to approximate the initial manifold is a hyperparameter that you can learn for different datasets.

The pretrained model is a lstm-VAE trained with KL-annealing. But for new datasets, you could also try more fancy models such as vmf-VAE [1], cnn-VAE [2], SA-VAE [3], or lagging-VAE [4]. Please let me know if you have some good results on new datasets. Thanks again!

ref:
[1]: https://arxiv.org/abs/1808.10805
[2]: https://arxiv.org/abs/1702.08139
[3]: https://arxiv.org/pdf/1802.02550.pdf
[4]: https://openreview.net/forum?id=rylDfnCqF7
